pkg/dockerregistry/server/signaturedispatcher.go

```diff
@@ -0,0 +1,147 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+
+	ctxu "github.com/docker/distribution/context"
+
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/handlers"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+
+	gorillahandlers "github.com/gorilla/handlers"
+)
+
+type Signature struct {
+	Name    string `json:"name"`
+	Type    string `json:"type"`
+	Content []byte `json:"content"`
+}
+
+type SignatureList struct {
+	Signatures []Signature `json:"signatures"`
+}
+
+const defaultSignatureType = "atomic"
+
+// SignatureDispatcher dispatch the signatures endpoint.
+func SignatureDispatcher(ctx *handlers.Context, r *http.Request) http.Handler {
+	signatureHandler := &signatureHandler{
+		Context:           ctx,
+		originalReference: ctxu.GetStringValue(ctx, "vars.name") + "@" + ctxu.GetStringValue(ctx, "vars.digest"),
+	}
+
+	signatureHandler.Reference, _ = imageapi.ParseDockerImageReference(signatureHandler.originalReference)
+
+	return gorillahandlers.MethodHandler{
+		"GET": http.HandlerFunc(signatureHandler.Get),
+		"PUT": http.HandlerFunc(signatureHandler.Put),
+	}
+}
+
+type signatureHandler struct {
+	Context           *handlers.Context
+	Reference         imageapi.DockerImageReference
+	originalReference string
+}
+
+func (s *signatureHandler) Put(w http.ResponseWriter, r *http.Request) {
+	context.GetLogger(s.Context).Debugf("(*signatureHandler).Create")
+	client, ok := UserClientFrom(s.Context)
+	if !ok {
+		s.handleError(s.Context, errcode.ErrorCodeUnknown.WithDetail("unable to get origin client"), w)
+		return
+	}
+
+	sig := Signature{}
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		s.handleError(s.Context, errcode.ErrorCodeUnknown.WithDetail(err.Error()), w)
+		return
+	}
+	if err := json.Unmarshal(body, &sig); err != nil {
+		s.handleError(s.Context, errcode.ErrorCodeUnknown.WithDetail(err.Error()), w)
+		return
+	}
+
+	if len(sig.Type) == 0 {
+		sig.Type = defaultSignatureType
+	}
+	newSig := &imageapi.ImageSignature{Content: sig.Content, Type: sig.Type}
+	newSig.Name = sig.Name
+	if _, err := client.ImageSignatures().Create(newSig); err != nil {
+		if kapierrors.IsUnauthorized(err) {
+			s.handleError(s.Context, errcode.ErrorCodeUnauthorized.WithDetail(fmt.Sprintf("not authorized to create image %q signature: %v", s.Reference.String(), err)), w)
+			return
+		}
+		if kapierrors.IsNotFound(err) {
```

Comment:
This could handle `BadRequest` and `AlreadyExists` as well. For the latter, you might need to register your own `ErrorCodeSignatureExists`.